lib/fastlane/actions/xcode_server_get_assets.rb

```diff
@@ -0,0 +1,299 @@
+
+        # now find the archive and unzip it
+        zipped_archive_path = asset_entries.select { |i| i.end_with?('xcarchive.zip') }.first
+
+        if zipped_archive_path
+
+          Helper.log.info "Found an archive in the assets folder...".yellow
+
+          archive_file_path = File.basename(zipped_archive_path, File.extname(zipped_archive_path))
+          archive_dir_path = File.dirname(zipped_archive_path)
+          archive_path = File.join(archive_dir_path, archive_file_path)
+          if File.exists?(archive_path)
+            # we already have the archive, skip
+            Helper.log.info "Archive #{archive_path} already exists, not unzipping again...".yellow
+          else
+            # unzip the archive
+            sh "unzip -q \"#{zipped_archive_path}\" -d \"#{archive_dir_path}\""
+          end
+
+          # reload asset entries to also contain the xcarchive file
+          asset_entries = Dir.entries(assets_path).map { |i| File.join(assets_path, i) }
+
+          # optionally delete everything except for the archive
+          if !keep_all_assets
+            files_to_delete = asset_entries.select do |i| 
+              File.extname(i) != '.xcarchive' && ![".", ".."].include?(File.basename(i))
+            end
+
+            files_to_delete.each do |i|
+              FileUtils.rm_rf(i)
+            end
+          end
+
+          Actions.lane_context[SharedValues::XCODE_SERVER_GET_ASSETS_ARCHIVE_PATH] = archive_path
+        end
+        
+        Actions.lane_context[SharedValues::XCODE_SERVER_GET_ASSETS_PATH] = assets_path
+
+        return assets_path
+      end
+
+      class XcodeServer
+
+        def initialize(host, username, password)
+          @host = host.start_with?('https://') ? host : "https://#{host}"
+          @username = username
+          @password = password
+        end
+
+        def fetch_all_bots
+          response = get_endpoint('/bots')
+          raise "You are unauthorized to access data on #{@host}, please check that you're passing in a correct username and password.".red if response.status == 401 
+          raise "Failed to fetch Bots from Xcode Server at #{@host}, response: #{response.status}: #{response.body}.".red if response.status != 200
+          bots = JSON.parse(response.body)['results']
+        end
+
+        def fetch_integrations(bot_id)
+          response = get_endpoint("/bots/#{bot_id}/integrations?limit=10")
+          raise "Failed to fetch Integrations for Bot #{bot_id} from Xcode Server at #{@host}, response: #{response.status}: #{response.body}".red if response.status != 200
+          integrations = JSON.parse(response.body)['results']
+        end
+
+        def fetch_assets(integration_id, target_folder, action)
+          
+          # create a temp folder and a file, stream the download into it
+          Dir.mktmpdir do |dir|
+
+            temp_file = File.join(dir, "tmp_download.#{rand(1000000)}")
```

Comment:
Yeah I guess, using `rand` feels good enough to me really though. Not sure what's more likely to clash.